### PR TITLE
csvjson: Support types other than Point

### PR DESCRIPTION
**Pull request by jayvdb**
Parses and removes 'geojson' from the properties,
putting the coordinate into the GeoJSON feature geometry.

Fixes https://github.com/wireservice/csvkit/issues/866

### CODE REVIEW

**Comment by jayvdb** [comment]
I'm not super happy about the `global`, and otherwise happy to revise patch as requested, or split it into several commits.

**Comment by jpmckinney** [comment]
Python 3 has `nonlocal`, but we can't use it since we need to support Python 2.

What if you make them class properties (`self.min_lon`, etc.)?

File: csvkit/utilities/csvjson.py [hunk 1]
```diff
@@ -110,39 +144,48 @@ def dump_json(data, newline=False):
...
                 properties = OrderedDict()
                 geoid = None
+                geo_type = 'Point'
                 lat = None
                 lon = None
+                coords = None
 
                 for i, c in enumerate(row):
                     if i == lat_column:
+                        if not c:
```
Comment (jpmckinney) [inline csvkit/utilities/csvjson.py]: I think the issue is that if `c` is `None`, then `float(c)` raises `TypeError` instead of `ValueError`. Can you make a separate PR that fixes the issue of `c` being `None`, by catching `TypeError`? It's separate from the rest of this issue.

File: csvkit/utilities/csvjson.py [hunk 2]
```diff
@@ -92,11 +92,45 @@ def dump_json(data, newline=False):
...
+                    max_lon = lon
+
+            def update_boundary_coords(coords):
+                if (isinstance(coords, list) and len(coords) == 2 and
+                        isinstance(coords[0], float) and
+                        isinstance(coords[1], float)):
+                    update_boundary_lon(coords[0])
+                    update_boundary_lat(coords[1])
+                else:
+                    try:
```
Comment (jpmckinney) [inline csvkit/utilities/csvjson.py]: Can you explain in what scenario the coords from the geojson will fail to iterate?

File: csvkit/utilities/csvjson.py (hunk 1 above)
Comment (jayvdb) [inline csvkit/utilities/csvjson.py]: Yes, correct.
However, this only occurs when the type is not `Point`, and fixing this exception only causes the tool to generate a .geojson which is broken in other ways.  Fixing only the exception doesnt make the tool better; in fact, I would say that the exception is a good thing, as it prevents people from believing that their input was correctly parsed.

But, happy to split to separate it to a separate commit - other improvements needed anyway.

**Comment by jayvdb** [comment]
> What if you make them class properties (self.min_lon, etc.)?

My main reason to avoid this is efficiency.  Any non 'local' name results in a lot of extra name resolution work for large files, and this computation of the bbox is IMO not always desirable (see https://github.com/wireservice/csvkit/issues/868) so I am reticent to make it also inefficient .

File: csvkit/utilities/csvjson.py (hunk 1 above)
Comment (jpmckinney) [inline csvkit/utilities/csvjson.py]: Ah, nevermind, I confused this with a different comment you made about lat/lon being empty.

**Comment by jpmckinney** [comment]
Do you have a real speed comparison?

**Comment by jayvdb** [comment]
Happy to provide it this evening (WIB) when I get back into this project.

File: csvkit/utilities/csvjson.py (hunk 2 above)
Comment (jayvdb) [inline csvkit/utilities/csvjson.py]: I didnt find any.  Removing.
(I did a quick check of types that I dont use using https://github.com/node-geojson/geojson-flatten/tree/master/test/fixture , resulting in https://github.com/wireservice/csvkit/issues/871 ).  If there are still some types which might cause an exception, we need exceptions here to catch and fix them.

File: csvkit/utilities/csvjson.py (hunk 1 above)
Comment (jpmckinney) [inline csvkit/utilities/csvjson.py]: I'm wondering if a '0' is read as an integer `0` instead of the string `'0'`, then we should check `if c is None` instead.

**Comment by jayvdb** [comment]
On real data for the bbox computation, using cProfile, `csvjson.py:119(update_boundary_coords)` is the 24th most significant component by cumulative time.  This amount of time is quite insignificant to the total time, however much of the time is spent in csv reading.

Sorting by total time instead, it is the 10th most significant component.

Ordered by number of calls, number 11-13 are:
- csvjson.py:119(update_boundary_coords) : 287562/1691
- csvjson.py:111(update_boundary_lon) : 285911
- csvjson.py:103(update_boundary_lat) : 285911

That is a lot of calls!

I do not have analysis of switching from local vars to instance variables, as that is going to need a more involved benchmark/profile rig to avoid the CSV read/JSON write phases. (c.f. https://github.com/wireservice/csvkit/issues/872 )  But as seen by the number of calls, this is not trivial due to the recursive nature, as using instance variables is going to require at least 2 * 2 * 285911 extra `getattr` for my dataset.  I am not sure how well that is optimised internally by cPython , but it could be optimised given that these are very tight loops without any other alterations to the `self` attribute dictionary.

**Comment by jayvdb** [comment]
Do you still want me to get rid of the `global`s?

**Comment by jpmckinney** [comment]
Considering reading/writing take up so much time (and are absolutely necessary to do anything with csvkit), I think it's unnecessary optimization to make that loop more efficient, at the cost of potentially hard-to-fix bugs in the future due to the use of globals. So, removing globals would be best.

**Comment by jayvdb** [comment]
ok
